internal/feeder/bastion/bastion_feeder.go

```diff
@@ -0,0 +1,230 @@
+	"golang.org/x/mod/sumdb/note"
+	"golang.org/x/net/http2"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog/v2"
+)
+
+type Config struct {
+	Addr            string
+	Prefix          string
+	Logs            []config.Log
+	BastionKey      ed25519.PrivateKey
+	WitnessVerifier note.Verifier
+}
+
+// FeedBastion talks to the bastion to receive checkpoints to be witnessed.
+// This function returns once the provided context is done.
+func FeedBastion(ctx context.Context, c Config, w feeder.Witness) error {
+	klog.Infof("My bastion backend ID: %064x", sha256.Sum256(c.BastionKey.Public().(ed25519.PublicKey)))
+	h := &addHandler{
+		w:           w,
+		logs:        make(map[string]config.Log),
+		witVerifier: c.WitnessVerifier,
+	}
+	for _, l := range c.Logs {
+		l := l
+		h.logs[l.ID] = l
+	}
+
+	return connectAndServe(ctx, c.Addr, h, c.BastionKey)
+}
+
+type addHandler struct {
+	w           feeder.Witness
+	logs        map[string]config.Log
+	witVerifier note.Verifier
+}
+
+func (a *addHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+	_, proof, cp, err := parseBody(r.Body)
+	if err != nil {
+		klog.V(1).Infof("invalid body: %v", err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	s := strings.SplitN(string(cp), "\n", 2)
+	if len(s) != 2 {
+		klog.V(1).Infof("invalid cp: %v", cp)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	logID := log.ID(s[0])
+	logCfg, ok := a.logs[logID]
+	if !ok {
+		klog.V(1).Infof("unknown log: %v", logID)
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+	signedCP, updateErr := a.w.Update(context.Background(), logID, cp, proof)
+	checkpoint, _, n, cpErr := log.ParseCheckpoint(signedCP, logCfg.Origin, a.witVerifier)
+
+	if updateErr != nil {
+		if sc := status.Code(err); sc == codes.FailedPrecondition {
+			// Invalid proof
+			klog.V(1).Infof("invalid proof: %v", err)
+			w.WriteHeader(http.StatusForbidden)
+			return
+		}
+		if sc := status.Code(err); sc == codes.AlreadyExists {
+			// old checkpoint is smaller than the latest the witness knows about
+			w.Header().Add("Content-Type", "text/x.tlog.size")
+			w.WriteHeader(http.StatusConflict)
+			if _, err := w.Write([]byte(fmt.Sprintf("%d\n", checkpoint.Size))); err != nil {
+				klog.V(1).Infof("Failed to write size response: %v", err)
+			}
+			return
+		}
+	}
+
+	if cpErr != nil {
+		klog.V(1).Infof("invalid checkpoint: %v", cpErr)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if _, err := w.Write([]byte(fmt.Sprintf("— %s %s\n", n.Sigs[0].Name, n.Sigs[0].Base64))); err != nil {
+		klog.V(1).Infof("Failed to write signature response: %v", err)
+	}
+}
+
+// parseBody reads the incoming request and parses into constituent parts.
+//
+// The request body MUST be a sequence of
+// - a previous size line,
+// - zero or more consistency proof lines,
+// - and an empty line,
+// - followed by a [checkpoint][].
+func parseBody(r io.Reader) (uint64, [][]byte, []byte, error) {
```

Comment:
Tests for this function would be beneficial.

Reply:
Good point, added